lib/pbench/test/unit/agent/test_tool_group.py

```diff
@@ -0,0 +1,147 @@
+
+
+class TestToolGroup:
+
+    pbench_run = "/mock/pbench_run"
+    group = "tool-group"
+    mock_tg_dir_name = Path(pbench_run, f"tools-v1-{group}")
+
+    def mock_resolve(self: Path, strict: bool = False):
+        """
+        Return a "full" fake file name
+
+        Args:
+            strict: Not used in the mock
+
+        Returns:
+            new Path with fake resolution
+        """
+        return Path("/mock/pbench-agent/")
+
+    def mock_is_dir(self):
+        """Return 'True' if Directory exist"""
+        return True
+
+    def mock_listdir(self, dir_path):
+        """Return 'Hostname' or 'ToolNames' based on directory path
+
+        Args:
+            dir_path: Directory path
+        """
+        if str(dir_path).endswith("hostname"):
+            return ["tool1", "tool2"]
+        else:
+            return ["hostname"]
+
+    def mock_read_text(self):
+        """Return content of File"""
+        return ""
+
+    def test_pbench_run_value(self):
+        """Tests if pbench run value exists"""
+        expected_error_msg = f"Cannot validate tool group, '{self.group}', 'pbench_run' environment variable missing"
+        with pytest.raises(BadToolGroup) as exc:
+            ToolGroup.verify_tool_group(self.group)
+        assert expected_error_msg in str(exc)
+
+    def test_pbench_run_dir_exist(self):
+        """Tests if pbench_run directory exist"""
+        expected_error_msg = f"Bad tool group, '{self.group}': directory {self.mock_tg_dir_name} does not exist"
+        with pytest.raises(BadToolGroup) as exc:
+            ToolGroup.verify_tool_group(self.group, self.pbench_run)
+        assert expected_error_msg in str(exc)
+
+    def test_target_dir_exist(self, monkeypatch):
+        """Test target directory exist with pbench run value as environment value"""
+        pbench_run = "/mock/environ_val/pbench_run"
+        monkeypatch.setenv("pbench_run", pbench_run)
+        mock_tg_dir_name = Path(pbench_run, f"tools-v1-{self.group}")
+        expected_error_msg = f"Bad tool group, '{self.group}': directory {mock_tg_dir_name} does not exist"
+        with pytest.raises(BadToolGroup) as exc:
+            ToolGroup.verify_tool_group(self.group)
+        assert expected_error_msg in str(exc)
+
+    def test_target_dir_exception(self, monkeypatch):
+        """Verify Target Directory Path"""
+
+        def mock_resolve(self, strict=True):
+            """Mocked the check to see if the path exists"""
+            return self.path
+
+        expected_error_msg = f"Bad tool group, '{self.group}': error resolving {self.mock_tg_dir_name} directory"
+        monkeypatch.setattr(Path, "resolve", mock_resolve)
+        with pytest.raises(BadToolGroup) as exc:
+            ToolGroup.verify_tool_group(self.group, self.pbench_run)
+        assert expected_error_msg in str(exc)
+
+    def test_target_dir_is_directory(self, monkeypatch):
+        """Verify target Directory as Directory"""
+
+        def mock_is_dir(self) -> bool:
```

Comment:
It's a bit confusing to have the outer `mock_is_dir` and an inner `mock_is_dir`; maybe rename this to `mock_is_not_dir`? 😆

Reply:
Sure :)

Reply:
Or, follow my suggestions to either rename the mocks or just use `lambda` expressions instead of defining functions.